manager/integration/tests/test_scheduling.py

```diff
@@ -0,0 +1,349 @@
+    assert len(volume["replicas"]) == 3
+    check_volume_data(volume, data)
+
+    cleanup(client, volume)
+
+
+def test_hard_anti_affinity_scheduling(client, volume_name):  # NOQA
+    """
+    Test that volumes with Hard Anti-Affinity work as expected.
+
+    With Hard Anti-Affinity, scheduling on nodes with existing replicas should
+    be forbidden, resulting in "Degraded" state.
+    """
+    volume = create_and_check_volume(client, volume_name)
+    host_id = get_self_host_id()
+    volume.attach(hostId=host_id)
+    volume = wait_for_volume_healthy(client, volume_name)
+    assert len(volume["replicas"]) == 3
+
+    data = write_volume_random_data(volume)
+    setting = client.by_id_setting(SETTING_REPLICA_HARD_ANTI_AFFINITY)
+    client.update(setting, value="true")
+    node = client.by_id_node(host_id)
+    client.update(node, allowScheduling=False)
+    replica_names = map(lambda replica: replica.name, volume["replicas"])
+    host_replica = get_host_replica(volume, host_id)
+
+    volume.replicaRemove(name=host_replica["name"])
+    wait_no_new_replica_ready(client, volume_name, replica_names)
+    volume = wait_for_volume_degraded(client, volume_name)
+    # We can't just count the replicas since there may be more than two due to
+    # the manager attempting to schedule more. We need to make sure there are
+    # the correct number of Healthy replicas instead.
+    assert sum([1 for replica in volume["replicas"] if replica["running"] and
+                replica["mode"] == "RW"])
+    check_volume_data(volume, data)
+
+    cleanup(client, volume)
+
+
+def test_hard_anti_affinity_detach(client, volume_name):  # NOQA
+    """
+    Test that volumes with Hard Anti-Affinity are still able to detach and
+    reattach to a node properly, even in degraded state.
+    """
+    volume = create_and_check_volume(client, volume_name)
+    host_id = get_self_host_id()
+    volume.attach(hostId=host_id)
+    volume = wait_for_volume_healthy(client, volume_name)
+    assert len(volume["replicas"]) == 3
+
+    data = write_volume_random_data(volume)
+    setting = client.by_id_setting(SETTING_REPLICA_HARD_ANTI_AFFINITY)
+    client.update(setting, value="true")
+    node = client.by_id_node(host_id)
+    client.update(node, allowScheduling=False)
+    replica_names = map(lambda replica: replica.name, volume["replicas"])
+    host_replica = get_host_replica(volume, host_id)
+
+    volume.replicaRemove(name=host_replica["name"])
+    wait_no_new_replica_ready(client, volume_name, replica_names)
+    volume = wait_for_volume_degraded(client, volume_name)
+    volume.detach()
+    volume = wait_for_volume_detached(client, volume_name)
+    assert len(volume["replicas"]) == 2
+
+    volume.attach(hostId=host_id)
+    volume = wait_for_volume_degraded(client, volume_name)
+    # Make sure we're still not getting another successful replica.
+    new_replica_running = False
+    for i in range(RETRY_COUNTS):
```

Comment:
This one would take long as well... Feel like we can introduce some states to indicates that we cannot schedule more replicas.

Reply:
See my thoughts in the previous comment.